### ISSUE TITLE
Some code-simplification suggests

### ISSUE THREAD

**Issue by ianozsvald** [jmcarpenter2/swifter#73]
Hello. Thanks for the cool library. I'm working on the second edition of my High Performance Python book and I'm going to include a section on Swifter. I'll also be demoing this in a couple of upcoming conference talks.

I'm taking a look through your code and I'll raise some suggestions if that's ok? For those you agree with I'd happily raise a PR.

https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L262
`samp` was a mystery until I realised it was shorthand for `sample`. I'd suggest using the 2 extra characters for readability and adding a comment explaining the intention of the sample (to save figuring out the code that follows).

https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L268
`assert` in production code is dangerous as `python -O` switches off assert checks: https://docs.python.org/3/tutorial/modules.html#compiled-python-files

I ran foul of this recently https://twitter.com/pwang/status/1157239710193991682 and I've since switched to using Bulwark and following the `if ... raise ...` approach as Peter suggests, I'm in the process of writing some advice docs given that I teach this in my public courses: https://github.com/ZaxR/bulwark/issues/46

Maybe something like the following works (assuming I've correctly understood the intent):

```
sample_result = sample.apply(...)
if not sample_result.equals(tmp_df):
    raise AssertionError("Attempt at applying the function to a sample has failed") 
else:
    return func(self._obj, *args, **kwds)
```

https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L283
`n_repeats = 3` generally I'd suggest using the CONSTANTS convention i.e. `N_REPEATS = 3`.

https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L272
Do we need the `:` after the exception? We could catch the exception with a `pass` and then I think  the following code would run just the same? If so, it'd be easier to parse as a 3rd person who is unfamiliar with the code.

https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L300
I think these two returns can be compressed to something like:

```
if ...
  returned_func = self._obj.progress_apply
else:
  returned_func = self._obj.apply

return returned_func(
    func,
    axis=axis,
    broadcast=broadcast,
    raw=raw,
    reduce=reduce,
    result_type=result_type,
    args=args,
    **kwds
)
```

**Comment by jmcarpenter2** [comment]
Hi @ianozsvald, 

Thanks so much for your interest in `swifter` and desire to include the library in your book. I greatly appreciate your suggestions for improving the library and am currently making the code changes you suggested. I will submit a PR for review shortly. The only comment you made that I don't totally understand is listed below. Could you elaborate further on what you mean? I think this `except` is necessary, but I also think I'm misunderstanding what you're suggesting.

> https://github.com/jmcarpenter2/swifter/blob/master/swifter/swifter.py#L272
> Do we need the `:` after the exception? We could catch the exception with a `pass` and then I think the following code would run just the same? If so, it'd be easier to parse as a 3rd person who is unfamiliar with the code.

I'm also planning to devote some time today to issues listed under the `Enhancement` tag today. Hopefully I am able to make good progress on one of those issues and can release a new version with added functionality soon.

Thank you,
Jason

**Comment by ianozsvald** [comment]
Hey Jason, thanks for acting so quickly! 

Re. my final suggestion - I see that that's unclear. Following the code you essentially have:
try: pandas apply, if success then return
except: the apply failed so we use Dask and then exit the function

My thought (and feel free to ignore it!) is that you could say:
```
except AttributeError,...:
    pass
# all the original code for Dask is now-left-dedented 
```

This doesn't change the flow of the execution, it just makes it clearer on first inspection that the try/except is going something that only applies in the middle block of the function (and it returns if successful). 

Partly my confusion about the latter part of the function was because it was long (it was >1 screen on github), now you've compressed the final bit of `self._obj.progress_apply` to one return statement I suspect it'll be easier to read. Feel free to ignore this point if you don't think it matters!

Cheers, Ian.

**Comment by jmcarpenter2** [comment]
Hey @ianozsvald,

Thanks again for all the helpful comments. I actually think that the final suggestion may make the code a bit harder to interpret, even though it will help with the fitting on one page issue. So I left that out of the PR in the end.

But I realized that some added clarity was needed as well, so I introduced a few extra comments explaining what is going on at various stages. The PR is now merged and released.

As a side note, I did make some progress on the groupby-apply over the weekend, but am having trouble making it work inside of swifter. For some reason my approach seems to speed things up in a jupyter notebook, but not when I run it as an embedded part of swifter. I will be debugging these issues in the coming days/weeks, but will not have time to work on the library this weekend.

Thanks so much for your suggestions and contribution to the library, I hope you do not hesitate to suggest any other improvements as you use swifter more in preparation for the book/conference talks.
-Jason
